### ISSUE TITLE
[context] do providers retain references to consuming components?

### ISSUE THREAD

**Issue by benjamind** [webcomponents-cg/community-protocols#21]
Moving [this conversation](https://github.com/webcomponents/community-protocols/issues/19#issuecomment-905135476) here.

> My question is still related about the Provider. Its lifecycle, how we should work when we're implementing one. I see @benjamind comment made on July 7th makes sense.

> What feels unclear is if we have to have the Provider to hang on to that context-request Event reference, and make it so that it uses it when it detects it's time to disconnect

**Comment by benjamind** [comment]
I think you have the right idea, though I think this is off-topic to the issue here.

It may be helpful to check out the lit-context implementation I've started [here](https://github.com/lit/lit/pull/1955). Especially this [ContextProvider](https://github.com/lit/lit/pull/1955/files#diff-53fa6c8f234974eb9d4a8421925e380d5b2a34741a7e134c9328f722513d8021R38) implementation.

Assuming you wish to create a Context Provider which can deliver the context more than once, i.e. when its context value changes, will have to hold onto the callback function which is provided in the ContextEvent for any context it is interested in satisfying, which _also_ has the `multiple` property in the event set to true (indicating its willing to receive the value more than once). It shouldn't retain the Event object, only the callback from within it. Whenever the context value in the provider changes (not properties on the context value, the value reference itself), the provider should call the callbacks it has collected again.

One thing that I think is wrong in your implementation here is that your provider is driving the `requestUpdate` call on the consuming component. This I think is violating an encapsulation rule. The context provider should know nothing about the rendering lifecycle of a context consuming component, it should merely provide the value in the callback, and leave updating to the consumer.

**Comment by renoirb** [comment]
> **renoirb**: What feels unclear is if we have to have the Provider to hang on to that **context-request** Event reference, and make it so that it uses it when it detects it's time to disconnect.  

(Added after)

Like [a "*Stateful Context Manager*"; knowing which ContextAPI components are currently live, are accepting updates, etc.](https://gist.github.com/renoirb/21e31aab8d4cbcebb24afede7c49e449#file-context-api-ts-L85-L147)

Not necessarily to do updates, like we see below with `requestUpdate` (yes it is violation — That is copy-pasta from experiments from a month ago). But more about an utility to help track/untrack that can be used to help orchestrate consumer/provier operations or lifecycles.

```ts
/**
 * In experiment above, this was called UpdatableHonk — Naming things are hard.
 * Pretty much an event with a target property.
 */
export interface IContextEventWithTarget<T extends UnknownContext> extends IContextEvent<T> {
  readonly target: EventTarget
}

export class StatefulContextManager {
  contexts = new Map<string, Set<IContextEventWithTarget<Context<unknown>>>>()

  // Partially inspired by Microsoft Fast Foundation listenerMap
  // packages/web-components/fast-foundation/src/utilities/match-media-stylesheet-behavior.ts
  private listenerMap = new WeakMap<EventTarget, ContextCallback<unknown>>()

  respondFor(name: string, data: unknown) {
    const contexts = this.contexts
    if (contexts) {
      if (contexts.has(name) === false) {
        throw new Error(`StatefulContextManager respondFor: There is no context on the name ${name}`)
      }
      const entries = contexts.get(name)
      for (const { context, target, ...rest } of entries) {
        const callback = this.listenerMap.has(target) ? this.listenerMap.get(target) : void 0
        const payload = data ?? context.initialValue
        callback(payload)
      }
    }
  }

  protected keepTrackContextRequest<T extends UnknownContext>(event: ContextEvent<T>) {
    const { context } = event
    const { name } = context
    const contexts = this.contexts
    const callback: ContextCallback<ContextType<T>> = (value, dispose) => {
      event.callback(value, dispose)
      ;(event.target as LitElement).requestUpdate()
    }
    // ... some way of calling the call back and updating
    const honk: IContextEventWithTarget<T> = {
      multiple: false,
      ...event,
      callback,
      context,
      target: event.target,
    }
    this.listenerMap.set(event.target, callback)
    if (contexts) {
      if (contexts.has(name)) {
        contexts.get(name).add(honk)
      } else {
        contexts.set(name, new Set([honk]))
      }
    }
    event.stopPropagation()
  }

  /**
   * So we can make the Provider to add the handler (remove below)
   * Making this Stateful Manager separate from the providers themselves.
   */
  addEventListenerTo(host: HTMLElement) {
    host.addEventListener('context-request', this.keepTrackContextRequest.bind(this))
  }
  removeEventListenerTo(host: HTMLElement) {
    host.addEventListener('context-request', this.keepTrackContextRequest.bind(this))
  }
}
```

> It shouldn't retain the Event object, only the callback from within it

Right. And that callback may have access to internal state.

**Comment by benjamind** [comment]
No a provider doesn't need to lookup which components are context-request friendly. This is actually why moving to the context object as a key for the context event request is valuable.

When a component needs a context value, it will fire the ContextEvent, in that event payload is an object that is the shared agreed upon identifier for context values of the type the requesting component wants. Therefore if a provider which can satisfy that request intercepts the ContextEvent, it can be reasonably confident that the component that issued the event is playing by the Context API rules and will do the right thing with the value provided to the callback function that was attached to the event.

So there's no need for a registry, but there is need for a shared agreed upon object that is used to key the context-request event.

**Comment by renoirb** [comment]
(After having read lit/lit#1955)

What I've been talking here feels similar to what lit/lit#1955 is doing. You have an abstract "*[Provider](https://github.com/lit/lit/blob/a14a74a0/packages/labs/context/src/lib/controllers/context-provider.ts#L20-L46)*" and "*[Consumer](https://github.com/lit/lit/blob/a14a74a0dc4140cbadfc7f69f67f9a1bd44ebb04/packages/labs/context/src/lib/controllers/context-consumer.ts#L37-L73)*" (which also cleans after itself). Also a "[ContextContainer](https://github.com/lit/lit/blob/a14a74a0/packages/labs/context/src/lib/controllers/context-consumer.ts#L37-L73)"  to handle telling that things has to be updated.  — Which is close to what I've called here ** StatefulContextManager**, hehehe, [like **benjamind/lit-context**](https://github.com/benjamind/lit-context/blob/main/src/lib/controllers/context-map-provider.ts#L53) :spock: or [*DeadWisdom*’s Gist](https://gist.github.com/DeadWisdom/62bff0179366fff137038f0bdf616e11), there's also [kcmr/context-provider](https://github.com/kcmr/context-provider) that looks similar too.

I hope my messy comments makes sense. In the links here, we might want an initial implementation that can be used as a starting point.

**Comment by renoirb** [comment]
> No a provider doesn't need to lookup which components are context-request friendly. This is actually why moving to the context object as a key for the context event request is valuable.

OK! Makes sense.

That's why you have in [benjamind/lit-context](https://github.com/benjamind/lit-context/blob/main/src/lib/controllers/context-map-provider.ts#L53) `provide` and the other to couple them. Nothing in the middle. No registry of sorts.

> So there's no need for a registry, but there is need for a shared agreed upon object that is used to key the context-request event.

OK!

**Comment by renoirb** [comment]
If we consider making a dependency less, not bound to a framework package that provides Context API, we might want to provide a ContextProvider or ContextContainer. At least the boilerplate of it.

When making a package, we would ask package users to use the baked container,provider,consume in a way that would need to be extendable.

We could either provide that boilerplate as a class mixin or as "state manager" the component would instantiate at connectedCallback, or constructor time. Probably we could do the same for context-provider or context-container. So we have two bits who knows how to tie themselves together.

That side might be as useful as a "*shared agreed upon object that is used to key the context-request event*" because it would be the perfect place to enforce the contract.

What would that package do for this purpose?

**Comment by renoirb** [comment]
Shoot, I missed today's meeting. I'll make sure it's [in my calendar](https://github.com/webcomponents-cg/docs-and-guides/blob/main/CONTRIBUTING.md#normative-andor-substantive-contributions) so I don't miss **BREAKOUT: Community Protocols** next time

**Comment by renoirb** [comment]
After more experimentation.

Keeping the callback reference for reuse makes sense.

What can we do if we have a good amount of Context callback in memory dor the same Context. But each owner of that context reference might have a discriminant of some sort.

Say it's a property with an ID. The ContextConsumer is instantiated at constructor, we can't get that ID from the context.  Or should we?

Ideas?

**Comment by justinfagnani** [comment]
I feel like this is answered in the description of Context Providers here: https://github.com/webcomponents-cg/community-protocols/blob/main/proposals/context.md#context-providers

To quote an answer I ave in #39 :

> providers do need to retain references to consumers if 1) the provider can update its context value, and 2) the `multiple` property of the request event is true. If both of those aren't true, then there's no need to hold a reference. If both are true, the dispose callback allows for severing the reference.

**Comment by justinfagnani** [comment]
In https://github.com/webcomponents-cg/community-protocols/pull/40 I clarify the language to say that a provider must not retain a reference to the callback if `subscribe` (the new name) is not true.

**Comment by Westbrook** [comment]
Looks like we've resolved this with #40. Feel free to open a new issue if there's more focused discussion that remains.
